File: ex2.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

float sp2and(float input1sp, float input2sp){
    return input1sp*input2sp;
}

float spNot(float input_sp){
    return 1.0f-input_sp;
}

float sa_from_sp(float s){
    return 2*s*(1.0f-s);
}

int getRandomNumber(double prob){
    return ((double)rand() / (double)RAND_MAX) < prob ? 1 : 0;
}

void monte_carlo(float pa, float pb, float pc, int n){
    int prev_a = getRandomNumber(pa);
    int prev_b = getRandomNumber(pb);
    int prev_c = getRandomNumber(pc);
    int prev_e = prev_a && prev_b;
    int prev_f = !prev_c;
    int prev_d = prev_e && prev_f;

    int count_e_ones = prev_e;
    int count_f_ones = prev_f;
    int count_d_ones = prev_d;

    int e_switches = 0;
    int f_switches = 0;
    int d_switches = 0;

    for(int i = 1; i < n; i++){
        int curr_a = getRandomNumber(pa);
        int curr_b = getRandomNumber(pb);
        int curr_c = getRandomNumber(pc);

        int curr_e = curr_a && curr_b;
        int curr_f = !curr_c;
        int curr_d = curr_e && curr_f;

        count_e_ones += curr_e;
        count_f_ones += curr_f;
        count_d_ones += curr_d;

        if(curr_e != prev_e) e_switches++;
        if(curr_f != prev_f) f_switches++;
        if(curr_d != prev_d) d_switches++;

        prev_a = curr_a;
        prev_b = curr_b;
        prev_c = curr_c;
        prev_e = curr_e;
        prev_f = curr_f;
        prev_d = curr_d;
    }

    float sp_e = (float)count_e_ones / n;
    float sp_f = (float)count_f_ones / n;
    float sp_d = (float)count_d_ones / n;

    float sa_e = (float)e_switches / (n - 1);
    float sa_f = (float)f_switches / (n - 1);
    float sa_d = (float)d_switches / (n - 1);

    printf("Monte Carlo for N=%d\n", n);
    printf("Signal probabilities: \te : %f\t f : %f\t d: %f\n", sp_e, sp_f, sp_d);
    printf("Switching activities: \te : %f\t f : %f\t d: %f\n\n", sa_e, sa_f, sa_d);
}

void print_truth_table(){
    int a,b,c;
    printf("Truth table\n");
    printf("A\tB\tC\tE\tF\tD");
    for(a=0;a<=1;a++)
    for(b=0;b<=1;b++)
    for(c=0;c<=1;c++)
    {
        printf("\n\n %d \t %d \t %d \t %d \t %d \t %d", a,b,c,(a&b),(!c),((a&b)&(!c)));
    }
    printf("\n");
}
int main(int argc ,char *argv[]){
    if (argc == 1){
        print_truth_table();
        return 0;
    }else if(argc != 4){
        printf("Incorrect input. Please enter probabilities for a,b,c e.g. ./a.out 0.2 0.3 0.4");
        return 1;
    }

    float a = atof(argv[1]);
    float b = atof(argv[2]);
    float c = atof(argv[3]);

    float e = sp2and(a,b);
    float f = spNot(c);
    float d = sp2and(e,f);

    float d_sa = sa_from_sp(d);
    float e_sa = sa_from_sp(e);
    float f_sa = sa_from_sp(f);

    printf("Propabilities Input\t");
    printf("a = %f\t b = %f\t c = %f\n\n",a,b,c);
    printf("Signal probabilities: \t");
    printf("e = %f\t f = %f\t d = %f\n",e,f,d);
    printf("Switching activities: \t");
    printf("e = %f\t f = %f\t d = %f\n\n\n",e_sa,f_sa,d_sa);

    srand((unsigned)time(NULL));
    monte_carlo(a,b,c,10);
    monte_carlo(a,b,c,100);
    monte_carlo(a,b,c,5228);
    monte_carlo(a,b,c,5384);
}

File: octave/ex2.m
function ex2(a, b, c)
  fprintf("Probabilities Input:\n");
  fprintf("a = %.3f\t b = %.3f\t c = %.3f\n\n", a, b, c);
  e = sp2AND(a, b);
  f = spNOT(c);
  d = sp2AND(e, f);
  fprintf("Signal probabilities:\t e = %.6f\t f = %.6f\t d = %.6f\n", e, f, d);
  fprintf("Switching activities:\t e = %.6f\t f = %.6f\t d = %.6f\n\n", ...
          switching_activity(e), switching_activity(f), switching_activity(d));

  simulations_array = [10, 100, 5228, 5384];
  for n = simulations_array
    monte_carlo(a, b, c, n);
  endfor
endfunction

function [e, f, d] = circuit_logic(a, b, c)
  e = a * b;
  f = 1 - c;
  d = e * f;
endfunction

% generates 1 with probability p or 0 with probability 1-p
function val = random_signal(p)
    val = rand() < p;
endfunction

function monte_carlo(pa, pb, pc, n)
  prev_a = random_signal(pa);
  prev_b = random_signal(pb);
  prev_c = random_signal(pc);
  [prev_e, prev_f, prev_d] = circuit_logic(prev_a, prev_b, prev_c);

  count_e_ones = prev_e;
  count_f_ones = prev_f;
  count_d_ones = prev_d;

  e_switches = 0;
  f_switches = 0;
  d_switches = 0;

  for i = 2:n
    curr_a = random_signal(pa);
    curr_b = random_signal(pb);
    curr_c = random_signal(pc);
    [curr_e, curr_f, curr_d] = circuit_logic(curr_a, curr_b, curr_c);

    count_e_ones += curr_e;
    count_f_ones += curr_f;
    count_d_ones += curr_d;

    e_switches += (curr_e ~= prev_e);
    f_switches += (curr_f ~= prev_f);
    d_switches += (curr_d ~= prev_d);

    prev_a = curr_a;
    prev_b = curr_b;
    prev_c = curr_c;
    prev_e = curr_e;
    prev_f = curr_f;
    prev_d = curr_d;
  endfor

  sp_e = count_e_ones / n;
  sp_f = count_f_ones / n;
  sp_d = count_d_ones / n;

  sa_e = e_switches / (n - 1);
  sa_f = f_switches / (n - 1);
  sa_d = d_switches / (n - 1);

  fprintf("Monte Carlo for N = %d\n", n);
  fprintf("Signal probabilities:\t e = %.6f\t f = %.6f\t d = %.6f\n", sp_e, sp_f, sp_d);
  fprintf("Switching activities:\t e = %.6f\t f = %.6f\t d = %.6f\n\n", sa_e, sa_f, sa_d);
endfunction

